README.md
# ScaRLib -- Scala Multi-Agent Deep Reinforcement Learning Framework.
ScaRLib is a Scala library for defining collaborative learning systems with many agents, namely: CMARL systems. 
In particular, this library offers:
- Centralized and decentralized learning modes
- Typed DSL used for defining multi-agent learning tasks
- Binding with state-of-the-art deep learning libraries (torch)
- Integration with Alchemist (a large-scale multi-agent simulator) and ScaFi (an aggregate programming language) to define typical scenarios in collective adaptive  system.

## ScaRLib submodules

![scarlib-modules](.img/scarlib-modules.png)

### ScaRLib Core
The module `scarlib-core` implements all the abstractions that model the CMARL domain. 
The key element is the `system`, it might be of two different types: 
    i) Centralized Training Decentralized Execution system (`CTDESystem`)
    ii) Decentralized Training Decentralized Execution system (`DTDESystem`).
Basically, the system, is a collection of agents that interact within a shared environment
and that are trained to optimize a global or local reward signal expressed by
a reward function. 
Through this definition, we have mentioned the remaining concepts of the CMARL domain,
therefore, to create an experiment, it is necessary to define six basic elements:

- Action space: the set of actions each agent can perform, it could be easily defined 
        extending the trait `Action`, for example:
  ```scala
  object ActionSpace {
    case object North extends Action
    case object South extends Action
    case object East extends Action
    case object West extends Action
  
    def toSeq() = Seq(North, South, East, West)
  }
  ```
  
- State: represents all the information an agent knows about the Environment at a certain time, 
    it must extend the trait `State`

- Reward function: defines how good is an action given the state in which the agent is
     ```scala
  class SimpleRewardFunction() extends RewardFunction {
        def compute(currentState: State, action: Action, newState: State): Double = ???
  }
  ```
  
- Environment: provides feedback to the agent in the form of rewards or penalties 
      for each action taken in a given state

- Dataset: the storage for the experience accumulated over the time by the agents. 
    The tool provides a simple buffered queue, if needed a user might implement his own dataset 
    extending the trait `ReplayBuffer`

- Agents: the number of agents involved in the experiment

Another pre-implemented component is the learning algorithm: the DQN. It approximates the Q-function used in the 
    Q-Learning algorithm with a Neural Network to deal with the explosion of the state space.
    As with all the ML algorithms there are some hyper-parameters we can tune to optimize the learning, 
    for that reason we provide a way to specify them in a single point:
```scala
  case class LearningConfiguration(
    epsilon: Decay[Double] = new ExponentialDecay(0.9, 0.1, 0.01),
    gamma: Double = 0.9,
    learningRate: Double = 0.0005,
    batchSize: Int = 32,
    updateEach: Int = 100,
    random: Random = new Random(1),
    dqnFactory: DQNAbstractFactory
  )
  ```


### Alchemist - Scafi
The module `alchemist-scafi` provides the bindings with two state-of-the-art tools, 
    namely: [Scafi](https://github.com/scafi/scafi) 
    and [Alchemist](http://alchemistsimulator.github.io/).
The integration of these two tools is a game-changer because it introduces 
    significant potential in ScaRLib: 
    i) Scafi enables the usage of the Aggregate Programming paradigms 
        to express collective behaviours for the agents
    ii) Alchemist enables the definition of large-scale sets of agents 
        in complex distributed systems (e.g., swarm robotics).

The definition of an experiment does not change significantly, only two elements are added:

- Alchemist simulation definition: basically it is a YAML file containing the description 
    of the alchemist environment, for example:
    ```yaml
    incarnation: scafi
    network-model:
        type: ConnectWithinDistance
        parameters: [0.5]
    deployments:
        type: Grid
        parameters: [-5,-5,5,5,0.25,0.25]
    programs:
        - program:
        - time-distribution: 1
          type: Event
          actions:
          - type: RunScafiProgram
            parameters: [program]
        - program: send
    ```
- Aggregate program: the Scafi program that express the aggregate logic. For example,
    if we want express the state as the distances from the neighbours:
    ```scala
      val state = foldhoodPlus(Seq.empty)(_ ++ _)(Set(nbrVector))
    ```
### DSL Core
The module `dsl-core` allows for agile and flexible creation of CMARL training systems.  
Using a system like Scala, creating a typed DSL allows for capturing errors during compilation, rather than waiting
    for the actual system runs to intercept simple configuration errors.
The exposed DSL is a simple facade to the abstractions shown in the `scarlib-core` module.
An example of DSL usage is the following:
```scala
val system = learningSystem {
    rewardFunction { new MyRewardFunction() } 
    actions { MyAction.all} // action supported by the agent
    dataset { ReplayBuffer[State, Action](10000) } // shared memory
    agents { 50 } // select the number of agent
    environment {
        // select a specific environment
        "it.unibo.scarlib.experiments.myEnvironment"
    }
}
```

## How to use it:
The tool is published on Maven. 
To integrate it into your own repository, you need to add (using Gradle):
```kotlin
implementation("io.github.davidedomini:scarlib-core:$version")
implementation("io.github.davidedomini:dsl-core:$version")
```

### Contributors

<a href="https://github.com/AlchemistSimulator/Alchemist/graphs/contributors">
	<img src="https://contributors-img.web.app/image?repo=ScaRLib-group/ScaRLib" />
</a>



CHANGELOG.md
## [3.1.3](https://github.com/ScaRLib-group/ScaRLib/compare/3.1.2...3.1.3) (2023-12-11)


### Documentation

* improve scaladoc ([0870554](https://github.com/ScaRLib-group/ScaRLib/commit/087055495f90699bb8384a4828ce7c8c3d815fc2))

## [3.1.2](https://github.com/ScaRLib-group/ScaRLib/compare/3.1.1...3.1.2) (2023-11-16)


### Bug Fixes

* **DSL:** fix agents instantiation ([2490106](https://github.com/ScaRLib-group/ScaRLib/commit/2490106cc7fa67de2b4782775be816a4c3ce6cb0))

## [3.1.1](https://github.com/ScaRLib-group/ScaRLib/compare/3.1.0...3.1.1) (2023-11-16)


### Bug Fixes

* **Learner:** remove useless method record ([1b59dbc](https://github.com/ScaRLib-group/ScaRLib/commit/1b59dbc2fc3c314c414882b4c5052725fa5c0973))
* **ReplayBuffer:** insert method now takes Experience as input ([7018216](https://github.com/ScaRLib-group/ScaRLib/commit/7018216be4c991e116fe7a09ab40655dae735584))


### General maintenance

* update README.md ([f019d5e](https://github.com/ScaRLib-group/ScaRLib/commit/f019d5ea0166058292c328d359cf1324a224a85c))


### Build and continuous integration

* change license name in publishing metadata ([3b93d16](https://github.com/ScaRLib-group/ScaRLib/commit/3b93d16fc271c0e17a881643964f0ebfcf517d28))

## [3.1.0](https://github.com/ScaRLib-group/ScaRLib/compare/3.0.0...3.1.0) (2023-11-13)


### Features

* add implicit configuration to environment method ([#29](https://github.com/ScaRLib-group/ScaRLib/issues/29)) ([d922f48](https://github.com/ScaRLib-group/ScaRLib/commit/d922f48feb7f1b28bb7ac326faae2957ba416bc1))

## [3.0.0](https://github.com/ScaRLib-group/ScaRLib/compare/2.0.1...3.0.0) (2023-11-09)


### ⚠ BREAKING CHANGES

* new dsl api (#28)

### Features

* new dsl api ([#28](https://github.com/ScaRLib-group/ScaRLib/issues/28)) ([78d1843](https://github.com/ScaRLib-group/ScaRLib/commit/78d18439ae7da92cf7fcc7c7d58d3f25ef98e529))

## [2.0.1](https://github.com/ScaRLib-group/ScaRLib/compare/2.0.0...2.0.1) (2023-11-08)


### Bug Fixes

* **ReplayBuffer:** fix generics ([110127c](https://github.com/ScaRLib-group/ScaRLib/commit/110127cd416d748821460eaf66ee9c1aa36989cc))

## [2.0.0](https://github.com/ScaRLib-group/ScaRLib/compare/1.6.5...2.0.0) (2023-11-06)


### ⚠ BREAKING CHANGES

* improvements

### Features

* improvements ([27c22d7](https://github.com/ScaRLib-group/ScaRLib/commit/27c22d794204f17c9785b645c094f57230ffa7da))


### General maintenance

* add dependencies management with gradle catalogs ([01eec18](https://github.com/ScaRLib-group/ScaRLib/commit/01eec180cf91fa7a34f770d8e9448c70005f64eb))
* update release command ([73fec4b](https://github.com/ScaRLib-group/ScaRLib/commit/73fec4b05ba50c0ac1093f663184ae9b9a89b531))
* update test job ([60ccd76](https://github.com/ScaRLib-group/ScaRLib/commit/60ccd7621fa2d54819194430c4ab26c43085fef1))


### Tests

* add decay test ([577796d](https://github.com/ScaRLib-group/ScaRLib/commit/577796df4e964e02d4af127f64409c2ba7e38065))

## [1.6.5](https://github.com/ScaRLib-group/ScaRLib/compare/1.6.4...1.6.5) (2023-04-16)


### Documentation

* add scaladoc ([0bfa1be](https://github.com/ScaRLib-group/ScaRLib/commit/0bfa1be8c002b4c37faea0b3a7a77bf0658f7590))


### Build and continuous integration

* update pipeline ([dc66105](https://github.com/ScaRLib-group/ScaRLib/commit/dc66105a6eaefe5f54082672adb31a0b7892d9ff))


### General maintenance

* add a contributor ([c9b78b9](https://github.com/ScaRLib-group/ScaRLib/commit/c9b78b9aa9055276faa9c03fad344e05c168cbc0))
* add issue templates ([58d00ac](https://github.com/ScaRLib-group/ScaRLib/commit/58d00acecea253b7473e9cea4c7153fdf07c9ce4))
* add licence note in all scala files ([cd8065c](https://github.com/ScaRLib-group/ScaRLib/commit/cd8065cc294a3b683ffac2fb7eef35744175a766))
* add scaladoc artifact generation ([1e921f1](https://github.com/ScaRLib-group/ScaRLib/commit/1e921f1081c679e4fecf7c7fadd9bf7f738e3e6d))
* update license in publish on central plugin ([8a69a1f](https://github.com/ScaRLib-group/ScaRLib/commit/8a69a1f613f2d8249c3b001949f4bf5d90dfccf3))
* update release configuration ([4fcd324](https://github.com/ScaRLib-group/ScaRLib/commit/4fcd324596cc77c7606a16eeecc6a57544d10d34))
* update the license ([419aa84](https://github.com/ScaRLib-group/ScaRLib/commit/419aa84a5de210c984ca4197c7c4f7764da745f7))

## [1.6.4](https://github.com/ScaRLib-group/ScaRLib/compare/1.6.3...1.6.4) (2023-04-15)


### Bug Fixes

* **DeepQLearning:** correctly load the model ([dccce57](https://github.com/ScaRLib-group/ScaRLib/commit/dccce57e19d962400a3f47594846f89b0e483091))

## [1.6.3](https://github.com/ScaRLib-group/ScaRLib/compare/1.6.2...1.6.3) (2023-03-21)


### Bug Fixes

* **core:** make snapshot path configurable by the user ([89c1805](https://github.com/ScaRLib-group/ScaRLib/commit/89c18051429e905bc862cc24b3d22e6531602b6b))

## [1.6.2](https://github.com/ScaRLib-group/ScaRLib/compare/1.6.1...1.6.2) (2023-03-16)


### Bug Fixes

* **DQNFactory:** fix generic factory ([88777c6](https://github.com/ScaRLib-group/ScaRLib/commit/88777c66f946da50c6210fe34dd2770c2370f5c9))

## [1.6.1](https://github.com/ScaRLib-group/ScaRLib/compare/1.6.0...1.6.1) (2023-03-16)


### Bug Fixes

* **DQNAbstractFactory:** implement createNN as a generic method ([f2c42ba](https://github.com/ScaRLib-group/ScaRLib/commit/f2c42ba646396ba576b91fa890915bde75af2689))

## [1.6.0](https://github.com/ScaRLib-group/ScaRLib/compare/1.5.0...1.6.0) (2023-03-15)


### Features

* complete alchemist-scafi integration ([ea4f222](https://github.com/ScaRLib-group/ScaRLib/commit/ea4f2228a3fd398f15e37797fdec1b888b517389))


### General maintenance

* resolve rebase conflicts ([b9d25c7](https://github.com/ScaRLib-group/ScaRLib/commit/b9d25c73b4d7107f9a215bb5ca2f4c4ca0487ea4))


### Dependency updates

* **deps:** update dependency it.unibo.alchemist:alchemist to v25.14.6 ([d5e6249](https://github.com/ScaRLib-group/ScaRLib/commit/d5e624987d9d985733297120bd60f087bad6a2dd))

vmas/src/main/resources/Cleaning.py
import math
import random
import typing
from typing import Dict, Callable, List

import torch

import Device
import wandb
from torch import Tensor, tensor

from vmas import render_interactively
from vmas.simulator.core import Agent, Landmark, Sphere, World, Entity
from vmas.simulator.heuristic_policy import BaseHeuristicPolicy
from vmas.simulator.scenario import BaseScenario
from vmas.simulator.sensors import Lidar
from vmas.simulator.utils import Color, X, Y, ScenarioUtils

if typing.TYPE_CHECKING:
    from vmas.simulator.rendering import Geom


def normalize(_value: torch.Tensor) -> torch.Tensor:
    _min = -100
    _max = 100
    return (_value - _min) / (_max - _min)


def normalize(_value: torch.Tensor, _min, _max) -> torch.Tensor:
    return (_value - _min) / (_max - _min)


class CleaningScenario(BaseScenario):
    def make_world(self, batch_dim: int, device: torch.device, **kwargs) -> World:
        self.wandb = kwargs.get("wandb", None)
        self.n_agents = kwargs.get("n_agents", 5)
        self.n_targets = kwargs.get("n_targets", 7)
        self.active_targets = torch.full((batch_dim, 1), self.n_targets, device=device)
        self.target_distance = kwargs.get("target_distance", 0.25)
        self._min_dist_between_entities = kwargs.get("min_dist_between_entities", 0.2)
        self._lidar_range = kwargs.get("lidar_range", 1.5)
        self.targets_pos = kwargs.get("targets_pos", [tensor([0, 0], device=Device.get())] * self.n_targets)
        self.n_targets_per_env = torch.full((batch_dim, 1), self.n_targets, device=device)

        self.min_collision_distance = 0.005
        self.agent_radius = 0.06
        self.target_radius = self.agent_radius

        self.viewer_zoom = 1
        self.target_color = Color.GREEN

        # Make world
        world = World(
            batch_dim,
            device,
            x_semidim=1,
            y_semidim=1,
            collision_force=500,
            substeps=2,
            drag=0.25,
        )
        # Add agents
        entity_filter_agents: Callable[[Entity], bool] = lambda e: e.name.startswith(
            "agent"
        )
        entity_filter_targets: Callable[[Entity], bool] = lambda e: e.name.startswith(
            "target"
        )
        for i in range(self.n_agents):
            # Constraint: all agents have same action range and multiplier
            agent = Agent(
                name=f"agent_{i}",
                collide=True,
                shape=Sphere(radius=self.agent_radius),
                sensors=[
                    # Lidar(
                    #     world,
                    #     angle_start=0.05,
                    #     angle_end=2 * torch.pi + 0.05,
                    #     n_rays=50,
                    #     max_range=self._lidar_range,
                    #     entity_filter=entity_filter_agents,
                    #     render_color=Color.BLUE,
                    # ),
                    Lidar(
                        world,
                        n_rays=50,
                        max_range=self._lidar_range,
                        entity_filter=entity_filter_targets,
                        render_color=Color.GREEN,
                    ),
                ],
            )
            agent.prev_distances = torch.zeros((batch_dim, 1), device=device)
            world.add_agent(agent)

        self._targets = []
        for i in range(self.n_targets):
            target = Landmark(
                name=f"target_{i}",
                collide=True,
                movable=False,
                shape=Sphere(radius=self.target_radius),
                color=self.target_color,
            )
            world.add_landmark(target)
            self._targets.append(target)
            # for j in range(batch_dim):
            #     target.set_pos(self.targets_pos[i], j)

        return world

    def reset_world_at(self, env_index: int = None):
        placable_entities = self._targets[: self.n_targets] + self.world.agents
        ScenarioUtils.spawn_entities_randomly(
            entities=placable_entities,
            world=self.world,
            env_index=env_index,
            min_dist_between_entities=self._min_dist_between_entities,
            x_bounds=(-self.world.x_semidim, self.world.x_semidim),
            y_bounds=(-self.world.y_semidim, self.world.y_semidim),
        )
        # for target in self._targets[self.n_targets:]:
        #     target.set_pos(self.get_outside_pos(env_index), batch_index=env_index)
        # for i, target in enumerate(self._targets):
        #     for j in range(self.world.batch_dim):
        #         target.set_pos(self.targets_pos[i], j)

    def respawn_targets(self, agent: Agent):
        targets = self._targets
        targets_positions = torch.stack([t.state.pos for t in targets], dim=0).to(Device.get())
        distances = []
        expanded_agent_pos = agent.state.pos.unsqueeze(0)
        distance = torch.norm(targets_positions - expanded_agent_pos, dim=-1).unsqueeze(-1).to(Device.get())# - self.agent_radius - self.target_radius
        for (i, target) in enumerate(targets):
            for j in range(self.world.batch_dim):
                if distance[i][j] <= self.target_distance:
                    target.set_pos(torch.tensor([-10000, -10000]).to(Device.get()), batch_index=j)
                    self.active_targets[j] -= 1
                    # bool_check = self.active_targets.flatten() == 0
                    # if bool_check.all(True):
                    #     self.done()
                    #     return

    def random_pos(self):
        x_coord = torch.full((self.world.batch_dim, 1), random.uniform(-self.world.x_semidim, self.world.x_semidim))
        y_coord = torch.full((self.world.batch_dim, 1), random.uniform(-self.world.y_semidim, self.world.y_semidim))
        tensor = torch.cat((x_coord, y_coord), dim=1)
        return tensor

    def reward(self, agent: Agent):
        # return self.reward_pos(agent)
        return self.reward_lidar(agent)

    def reward_lidar(self, agent: Agent):
        k = 1
        targets_lidar = agent.sensors[0].measure()
        #agents_lidar = agent.sensors[0].measure()
        # get distances from nearest target
        # targets_positions = torch.stack([t.state.pos for t in self._targets], dim=0).to(Device.get())
        # min_distances_from_targets = (torch.norm(targets_positions - agent.state.pos.unsqueeze(0), dim=-1).unsqueeze(-1)
        #                               .to(Device.get()))
        # t = min_distances_from_targets.min(dim=0).values.float().squeeze(0).view(self.world.batch_dim, 1)
        min_distances_from_lidar = targets_lidar.min(dim=1, keepdim=True).values.float()
        final_rewards = min_distances_from_lidar.clone().to(Device.get())
        mask = min_distances_from_lidar >= self._lidar_range
        if mask.any():
            #rewards_for_agents_outside_lidar_range = -torch.exp(-t[mask]).to(Device.get())
            final_rewards[mask] = -1 #-(2/math.pi) * (torch.atan(k * t[mask]).to(Device.get())) #rewards_for_agents_outside_lidar_range  # -(t[~mask]**2)
        mask = min_distances_from_lidar <= self.target_distance
        if mask.any():
            final_rewards[mask] = torch.full([self.world.batch_dim, 1], self.n_targets, device=Device.get(), dtype=torch.float32) - (self.active_targets) + 1
        mask = (min_distances_from_lidar > self.target_distance) & (min_distances_from_lidar < self._lidar_range) & (min_distances_from_lidar < agent.prev_distances)
        if mask.any():
                final_rewards[mask] = -0.5 + ((-min_distances_from_lidar[mask] / self._lidar_range + 1) * -0.5 / 1) #1 - (min_distances_from_lidar[mask] / self._lidar_range)
        mask = (min_distances_from_lidar > self.target_distance) & (min_distances_from_lidar < self._lidar_range) & (min_distances_from_lidar >= agent.prev_distances)
        if mask.any():
            final_rewards[mask] = -1 + ((-min_distances_from_lidar[mask] / self._lidar_range + 1) * -0.5)
        #rewards_for_agents_inside_lidar_range = -torch.exp(-t[mask]).to(Device.get())/2#torch.sigmoid(temp[~new_mask]).to(Device.get())
            #final_rewards[mask] = -(2/math.pi) * (torch.atan(k * t[mask]).to(Device.get()))#-torch.exp(-t[mask]).to(Device.get())/2 #rewards_for_agents_inside_lidar_range
            #final_rewards[mask] = -1 + (t[mask].clone().to(Device.get()) - self._lidar_range) * (1 - 0) / (0 - self._lidar_range)

        # reward for agents
        # min_distances = agents_lidar.min(dim=1, keepdim=True).values.float()
        # mask = min_distances < self._lidar_range
        # temp = min_distances.clone()
        # temp[~mask] = -self._lidar_range/3.0
        # temp[~mask] /= min_distances[~mask]
        # min_distances[~mask] = temp[~mask]
        # temp = min_distances.float().clone()
        # temp[mask] = -self._lidar_range/3.0
        # temp[mask] /= min_distances[mask]
        # agents_reward = temp

        self.respawn_targets(agent)
        # print(agent.name)
        # print(targets_reward)
        # print(agents_reward)
        # print(final_reward)
        for i in range(self.world.batch_dim):
            if self.active_targets[i] == 0:
                final_rewards[i] = 10
        agent.prev_distances = min_distances_from_lidar
        return final_rewards

    def get_outside_pos(self, env_index):
        return torch.empty(
            (1, self.world.dim_p)
            if env_index is not None
            else (self.world.batch_dim, self.world.dim_p),
            device=self.world.device,
        ).uniform_(-1000 * self.world.x_semidim, -1000 * self.world.x_semidim)

    def agent_reward(self, agent):
        agent_index = self.world.agents.index(agent)

        agent.covering_reward[:] = 0
        targets_covered_by_agent = (
                self.agents_targets_dists[:, agent_index] < self._covering_range
        )
        num_covered_targets_covered_by_agent = (
                targets_covered_by_agent * self.covered_targets
        ).sum(dim=-1)
        agent.covering_reward += (
                num_covered_targets_covered_by_agent * self.covering_rew_coeff
        )
        return agent.covering_reward

    def observation(self, agent: Agent):
        lidar_1_measures = agent.sensors[0].measure()
        #lidar_2_measures = agent.sensors[1].measure()
        return torch.cat(
            [
                #agent.state.pos,  # 2
                #agent.state.vel,  # 2
                lidar_1_measures,  # 15
                #lidar_2_measures,  # 15
            ],
            dim=-1,
        ).to(Device.get())

    def info(self, agent: Agent) -> Dict[str, Tensor]:
        info = {
            "active_targets": self.active_targets
        }
        return info

    def done(self):
        return torch.zeros((self.world.batch_dim, 1), device=self.world.device)

    def extra_render(self, env_index: int = 0) -> "List[Geom]":
        from vmas.simulator import rendering

        geoms: List[Geom] = []
        # Target ranges
        for i, target in enumerate(self._targets):
            range_circle = rendering.make_circle(self.target_distance, filled=False)
            xform = rendering.Transform()
            xform.set_translation(*target.state.pos[env_index])
            range_circle.add_attr(xform)
            range_circle.set_color(*self.target_color.value)
            geoms.append(range_circle)
        # Communication lines
        for i, agent1 in enumerate(self.world.agents):
            for j, agent2 in enumerate(self.world.agents):
                if j <= i:
                    continue
                agent_dist = torch.linalg.vector_norm(
                    agent1.state.pos - agent2.state.pos, dim=-1
                )
                if agent_dist[env_index] <= self._comms_range:
                    color = Color.BLACK.value
                    line = rendering.Line(
                        (agent1.state.pos[env_index]),
                        (agent2.state.pos[env_index]),
                        width=1,
                    )
                    xform = rendering.Transform()
                    line.add_attr(xform)
                    line.set_color(*color)
                    geoms.append(line)

        return geoms
